Pico consensus

viquezclaudio opened the pull request:
Pico Consensus Implementation
    
The PicoMacroSync is a macro sync mechanism that pushes the latest election block
that is obtained from a peer, if subsequent (sufficient) peers also have the same latest election
(and latest checkpoint blocks) then we move all the peers to live sync and we establish consensus.
If some peer has a different state than the initial state we obtained, we need to fallback to the
regular light macro sync.




To use it or try it in the web client:
web-client/src/client/lib.rs

`
let mut config = builder.volatile().pico().build().expect("Build configuration failed");
`


You can also set consensus sync mode to 'pico' if you want to test it in a light client


## Pull request checklist

- [ ] All tests pass. The project builds and runs.
- [ ] I have resolved any merge conflicts.
- [ ] I have resolved all `clippy` and `rustfmt` warnings.

viquezclaudio committed:
Pico Consensus Implementation

The PicoMacroSync is a macro sync mechanism that pushes the latest election block
that is obtained from a peer, if subsequent (sufficient) peers also have the same latest election
(and latest checkpoint blocks) then we move all the peers to live sync and we establish consensus.
If some peer has a different state than the initial state we obtained, we need to fallback to the
regular light macro sync.

jsdanielh commented on light-blockchain/src/blockchain.rs:
```diff
@@ -79,6 +80,23 @@ impl LightBlockchain {
         }
     }
 
+    // Resets the light blockchain to its initial state
```
```suggestion
    /// Resets the light blockchain to its initial state
```

jsdanielh commented on lib/src/config/config_file/mod.rs:
```diff
@@ -271,6 +271,9 @@ pub enum SyncMode {
     /// Light nodes use LightMacroSync + BlockLiveSync to reach consensus
     Light,
+    /// Pico nodes: They use PicoMacroSync optimistically
+    ///  if something fails, they fallback to LightMacroSync
```
```suggestion
    /// if something fails, they fallback to LightMacroSync
```

jsdanielh commented on lib/src/config/config.rs:
```diff
@@ -59,6 +59,9 @@ pub enum SyncMode {
     /// Light nodes: They use LightMacroSync + BlockLiveSync
     Light,
+    /// Pico nodes: They use PicoMacroSync optimistically
+    ///  if something fails, they fallback to LightMacroSync
```
```suggestion
    /// if something fails, they fallback to LightMacroSync
```

jsdanielh commented on light-blockchain/src/sync.rs:
```diff
@@ -190,4 +191,71 @@ impl LightBlockchain {
         Ok(PushResult::Extended)
     }
+
+    // Pushes an election block for the pico sync, bypassing many checks...
```
```suggestion
    /// Pushes an election block for the pico sync, bypassing many checks
```

jsdanielh commented on consensus/src/sync/light/mod.rs:
```diff
@@ -4,4 +4,4 @@ mod sync_stream;
 mod validity_window;
 
-pub use sync::LightMacroSync;
+pub use sync::{EpochIds, LightMacroSync, PeerMacroRequests};
```
Why do we need to make it public?

viquezclaudio replied:
Because they need to be accesible from the pico variant, otherwise:

```
rror[E0432]: unresolved imports `crate::sync::light::EpochIds`, `crate::sync::light::PeerMacroRequests`
  --> consensus/src/sync/pico/sync_requests.rs:17:17
   |
17 |         light::{EpochIds, PeerMacroRequests},
   |                 ^^^^^^^^  ^^^^^^^^^^^^^^^^^ no `PeerMacroRequests` in `sync::light`
   |                 |
   |                 no `EpochIds` in `sync::light`
   |
note: these structs exist but are inaccessible
```

viquezclaudio replied:
Another option would be to move those structs to some shared file between light and macro sync

jsdanielh replied:
Yeah, to me it means they maybe don't belong here

jsdanielh commented on consensus/src/sync/light/sync.rs:
```diff
@@ -34,7 +32,7 @@ use crate::{
 /// This struct is used to request Epochs IDs (hashes) from other peers
 /// in order to determine their macro chain state relative to us
-pub(crate) struct EpochIds<T> {
+pub struct EpochIds<T> {
```
Why do we need to make it public?

viquezclaudio replied:
Same as above

jsdanielh replied:
But the whole crate should be able to access it

jsdanielh commented on consensus/src/sync/light/sync.rs:
```diff
@@ -254,6 +257,7 @@ impl<TNetwork: Network> LightMacroSync<TNetwork> {
     }
 
     pub fn remove_peer_requests(&mut self, peer_id: TNetwork::PeerId) {
+        self.syncing_peers.remove(&peer_id);
```
I believe semantically, this function now maybe needs to be called `remove_peer` instead?

viquezclaudio replied:
Syncing peers is a HashSet, so the 'remove' function comes from the HashSet function name.
To change the name I would need to create a wrapper just for this purpose, and I dont think thats worth it

jsdanielh commented on consensus/src/sync/pico/sync.rs:
```diff
@@ -0,0 +1,116 @@
+    >,
+    /// Collection of peers we are currently pico macro syncing with
+    pub(crate) syncing_peers: HashSet<TNetwork::PeerId>,
+    /// Reference to the ZKP proxy used to interact with the ZKP componen
```
```suggestion
    /// Reference to the ZKP proxy used to interact with the ZKP component
```

jsdanielh commented on consensus/src/sync/pico/sync.rs:
```diff
@@ -0,0 +1,116 @@
+    /// Collection of peers we are currently pico macro syncing with
+    pub(crate) syncing_peers: HashSet<TNetwork::PeerId>,
+    /// Reference to the ZKP proxy used to interact with the ZKP componen
+    pub(crate) zkp_component_proxy: ZKPComponentProxy<TNetwork>,
```
Do we need the it for pico sync?

viquezclaudio replied:
Not for the pico sync, but when we fallback to the light macro sync we need to supply this reference

jsdanielh replied:
Could you add it as part of the `rustdoc`?

jsdanielh commented on consensus/src/sync/pico/sync.rs:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    pub fn remove_peer_requests(&mut self, peer_id: TNetwork::PeerId) {
```
Could you add some rustdoc?

jsdanielh commented on consensus/src/sync/pico/sync.rs:
```diff
@@ -0,0 +1,116 @@
+        self.syncing_peers.remove(&peer_id);
+    }
+
+    pub fn disconnect_peer(&mut self, peer_id: TNetwork::PeerId, reason: CloseReason) {
```
Could you add some rustdoc?

jsdanielh commented on consensus/src/sync/light/sync.rs:
```diff
@@ -275,8 +279,16 @@ impl<TNetwork: Network> MacroSync<TNetwork::PeerId> for LightMacroSync<TNetwork>
+    }
+
+    fn peers(&mut self) -> Vec<TNetwork::PeerId> {
+        self.syncing_peers.drain().collect()
```
With drain we're clearing the set. Why?

viquezclaudio replied:
Because when we call this function we want to obtain the peers to fallback to the other macro sync mechanism, so we dont want or need to keep any peer inside

jsdanielh replied:
So maybe this function should be name `drain_peers` instead

jsdanielh commented on consensus/src/sync/pico/sync.rs:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    fn peers(&mut self) -> Vec<TNetwork::PeerId> {
+        self.syncing_peers.drain().collect()
```
With drain we're clearing the set, why?

viquezclaudio replied:
Because when we call this function we want to obtain the peers to fallback to the other macro sync mechanism, so we dont want or need to keep any peer inside

jsdanielh replied:
So maybe this function should be name `drain_peers` instead. Right now from the name it looks like a getter but it's not

jsdanielh commented on consensus/src/sync/pico/sync_requests.rs:
```diff
@@ -0,0 +1,271 @@
+};
+
+impl<TNetwork: Network> PicoMacroSync<TNetwork> {
+    pub(crate) async fn request_epoch_ids(
```
Could you add some `rustdoc`?

jsdanielh commented on consensus/src/sync/pico/sync_requests.rs:
```diff
@@ -0,0 +1,271 @@
+        }
+    }
+
+    pub(crate) fn request_macro_headers(
```
Could you add some `rustdoc`?

jsdanielh commented on consensus/src/sync/pico/sync_requests.rs:
```diff
@@ -0,0 +1,271 @@
+                return Some(epoch_ids.sender);
+            } else {
+                // Check that the epoch_id sent by the peer at our current epoch number corresponds to
+                // our accepted state. If it doesn't, the peer is on a "permanent" fork, so we ban it.
```
Are we banning it?

viquezclaudio replied:
No, we cannot ban it, because we dont know if the peer is outdated or if it is being malicious.
We can only know this once we fallback to light macro sync

jsdanielh replied:
So maybe we need to update the comment?

jsdanielh commented on consensus/src/sync/pico/sync_requests.rs:
```diff
@@ -0,0 +1,271 @@
+        None
+    }
+
+    pub async fn request_macro_block(
```
Could you add some `rustdoc`?

jsdanielh commented on consensus/src/sync/pico/sync_requests.rs:
```diff
@@ -0,0 +1,271 @@
+            .await
+    }
+
+    pub async fn request_macro_chain(
```
Could you add some `rustdoc`?

jsdanielh commented on consensus/src/sync/pico/sync_stream.rs:
```diff
@@ -0,0 +1,253 @@
+                                                    block.clone(),
+                                                )
+                                            } else {
+                                                // We dont accept other election blocks if we are not in the genesis.
```
```suggestion
                                                // We don't accept other election blocks if we are not in the genesis.
```

jsdanielh commented on consensus/src/sync/pico/sync_stream.rs:
```diff
@@ -0,0 +1,253 @@
+                                    }
+                                    BlockchainProxy::Light(ref light_blockchain) => {
+                                        let blockchain = light_blockchain.upgradable_read();
+
```
```suggestion
```

jsdanielh commented on consensus/src/sync/pico/sync_stream.rs:
```diff
@@ -0,0 +1,253 @@
+                                                ));
+                                            }
+                                        } else {
+                                            // We only accept macro blocks from the current epoch
```
Shouldn't we check this is a macro block?

jsdanielh commented on consensus/src/sync/pico/sync_stream.rs:
```diff
@@ -0,0 +1,253 @@
+                                            %peer_id,
+                                            "Failed to push macro block",
+                                        );
+                                        // Since we cannot differentiate between a malicious peer and someone whos behind and
```
```suggestion
                                        // Since we cannot differentiate between a malicious peer and someone who's behind and
```

jsdanielh commented on consensus/src/sync/pico/sync_stream.rs:
```diff
@@ -0,0 +1,253 @@
+                                            "Failed to push macro block",
+                                        );
+                                        // Since we cannot differentiate between a malicious peer and someone whos behind and
+                                        // we dont know if the initial state (election block) that we obtained was good
```
```suggestion
                                        // we don't know if the initial state (election block) that we obtained was good
```

jsdanielh commented on consensus/src/sync/pico/sync_stream.rs:
```diff
@@ -0,0 +1,253 @@
+                                    }
+                                }
+                            }
+                            //  At this point we applied all the pending requests from this peer
```
```suggestion
                            // At this point we applied all the pending requests from this peer
```

jsdanielh commented on consensus/src/sync/syncer.rs:
```diff
@@ -324,7 +331,15 @@ impl<N: Network, M: MacroSync<N::PeerId>, L: LiveSync<N>> Stream for Syncer<N, M
+
+                    self.macro_sync.fallback(syncing_peers);
+                }
+                _ => {}
```
Isn't it better to keep the `None` to make the compiler alert us if we add another `enum` variant?

jsdanielh commented:
Could you also consider adding a pico node to the devnet scenarios? I think it would be valuable to test it as well

viquezclaudio committed:
Move common sync functionality

Create a sync_interface file where common functionality from the different
sync implementations can be imported

viquezclaudio committed:
Unittests for pico consensus

jsdanielh commented on .github/devnet_topologies/validators_stress.toml:
```diff
@@ -42,3 +42,7 @@ sync_mode = "history"
+
+[[node]]
+restartable = false
+sync_mode = "pico"
```
I think you also need to add it to at least [here](https://github.com/nimiq/core-rs-albatross/blob/906088071bafb69266cd33d01602e69e0810d0bd/scripts/devnet/topology.py#L108).

viquezclaudio committed:
Fix issue when theres nothing new to learn from a peer

When processing epoch ids it could happen that by the time we requested the epoch ids
and the time we received those we no longer have anything new to learn from the peer,
so we emit it as good.

viquezclaudio committed:
Check we receive macro blocks during macro sync

- Add a check that we receive macro blocks while we are macro syncing
- Rename collect_peers to drain_peers as per PR review

jsdanielh reviewed (approved):
LGTM
